src/arch/arm/kernel/boot.c

```diff
@@ -79,9 +79,14 @@ BOOT_CODE static bool_t arch_init_freemem(p_region_t ui_p_reg,
             reserved[index].end = ui_reg.end;
             index++;
         }
-    } else if (MODE_RESERVED == 1) {
-        reserved[index] = mode_reserved_region[0];
-        index++;
+    } else {
+        if (MODE_RESERVED == 1) {
+            reserved[index] = mode_reserved_region[0];
+            index++;
+        }
+
+        /* Reserve the ui_p_reg region still so it doesn't get turned into device UT. */
+        reserve_region(ui_p_reg);
```

Comment:
We are not doing the check for `ui_p_reg.start < PADDR_TOP` in RISC-V, so maybe we should remove it here also? Wouldn't `reserve_region()` then be calles also?

Reply:
This is necessary on aarch32 but not riscv32 as far as I could tell.  (It's not an issue on 64bit because the physical space can easily fit in 64-bit pointer width).

Because aarch32 has platforms with seL4_UserTop high enough that the kernel window is too small to fit all of the platform's available RAM, when the available RAM region is converted to a pptr it's `.end` value can be wrapped around and is less than it's `.start` value.  Because the reserved region_t array given to `init_freemem` must have the regions ordered from lowest to highest in the pptr range, this overflow means a region that is higher in RAM becomes lower in the reserved list.  This made the arch_freemem implementation on Arm more complicated [1] and it ends up being much simpler to reserve directly like how kernel devices are reserved because we know that it can't become regular untypeds already.  This isn't an issue for riscv32 at the moment because there aren't any platforms with Ram outside of the kernel window and so there can't be any wraparound for the user image region. 

[1]
```
diff --git a/src/arch/arm/kernel/boot.c b/src/arch/arm/kernel/boot.c
index 65d30a12..0565193f 100644
--- a/src/arch/arm/kernel/boot.c
+++ b/src/arch/arm/kernel/boot.c
@@ -42,10 +42,17 @@ BOOT_CODE static bool_t arch_init_freemem(p_region_t ui_p_reg,
                                           v_region_t it_v_reg,
                                           word_t extra_bi_size_bits)
 {
-    reserved[0].start = KERNEL_ELF_BASE;
-    reserved[0].end = (pptr_t)ki_end;
-
-    int index = 1;
+    int index = 0;
+    region_t ui_reg = paddr_to_pptr_reg(ui_p_reg);
+    int ui_reg_placed = 0;
+    if (KERNEL_ELF_BASE > ui_reg.start) {
+        reserved[index] = ui_reg;
+        index++;
+        ui_reg_placed = 1;
+    }
+    reserved[index].start = KERNEL_ELF_BASE;
+    reserved[index].end = (pptr_t)ki_end;
+    index++;
     if (dtb_p_reg.start) {
         /* the dtb region could be empty */
         reserved[index].start = (pptr_t) paddr_to_pptr(dtb_p_reg.start);
@@ -58,29 +65,32 @@ BOOT_CODE static bool_t arch_init_freemem(p_region_t ui_p_reg,
         return false;
     }
 
-    if (ui_p_reg.start < PADDR_TOP) {
-        region_t ui_reg = paddr_to_pptr_reg(ui_p_reg);
-        if (MODE_RESERVED == 1) {
-            if (ui_reg.end > mode_reserved_region[0].start) {
+    if (MODE_RESERVED == 1) {
+        if (!ui_reg_placed) {
+            if (ui_reg.start > mode_reserved_region[0].start) {
                 reserved[index] = mode_reserved_region[0];
                 index++;
                 reserved[index].start = ui_reg.start;
                 reserved[index].end = ui_reg.end;
+                index++;
+
             } else {
                 reserved[index].start = ui_reg.start;
                 reserved[index].end = ui_reg.end;
                 index++;
                 reserved[index] = mode_reserved_region[0];
+                index++;
             }
-            index++;
         } else {
-            reserved[index].start = ui_reg.start;
-            reserved[index].end = ui_reg.end;
+            reserved[index] = mode_reserved_region[0];
             index++;
         }
-    } else if (MODE_RESERVED == 1) {
-        reserved[index] = mode_reserved_region[0];
+    }
+
+    if (!ui_reg_placed) {
+        reserved[index] = ui_reg;
         index++;
+        ui_reg_placed = 1;
     }
 
     return init_freemem(get_num_avail_p_regs(), get_avail_p_regs(), index,
```

Reply:
But for the sake of unifying the general behavior, should this be done in the generic RISC-V also? Just because there is no platform now it a poor argument, it could happen when testing/exploring something new and then we might see random crashes that are a pain to investigate.

Reply:
> But for the sake of unifying the general behavior, should this be done in the generic RISC-V also? Just because there is no platform now it a poor argument, it could happen when testing/exploring something new and then we might see random crashes that are a pain to investigate.

This change shouldn't also be done for the RISC-V arch implementation of  arch_init_freemem.  The implementations are already arch specific.  RISC-V unconditionally adds the `ui_reg` region to the reserved region list unconditionally and so doesn't have the same issue on Arm where additional UT would be created.  

Whether RISC-V should be unconditionally adding the `ui_reg` to the reserved region isn't a concern of this PR. If some platform was added on RISC-V 32bit in the future that loaded the `ui_reg` above the kernel window, it would cause the boot code to print an error and abort during boot in the consistency checks that init_freemem performs.